File: src/events/push.rs
use crate::errors::Error;
use crate::events::Event;
use crate::events::base::{PushCommit, Repository, User, WebhookMessage};
use regex::Regex;
use serde::Deserialize;

#[derive(Deserialize)]
pub struct PushEvent {
    pub commits: Vec<PushCommit>,
    #[serde(rename = "ref")]
    pub ref_: String,
    pub pusher: User,
    pub sender: User,
    pub repository: Repository,
}

impl Event for PushEvent {
    fn handle(&self) -> Result<Option<WebhookMessage>, Error> {
        if self.commits.len() == 0 {
            return Ok(None);
        }

        let link_re = Regex::new(r"\[([^]]+)]\((https?://[^)]+)\)")?;
        let md_re = Regex::new(r"(?m)^\s*#{1,3}\s+")?;
        let mut commits = String::new();

        for c in &self.commits {
            let mut lines = c.message.lines();
            let first = lines.next().unwrap_or("");
            let commit_msg = if lines.next().is_some() {
                format!("{first}...")
            } else {
                first.to_string()
            };

            let clean_link_msg = link_re
                .replace_all(&commit_msg, |caps: &regex::Captures| {
                    format!("[{}](<{}>)", &caps[1], &caps[2])
                })
                .to_string();
            let clean_md_msg = md_re.replace_all(&clean_link_msg, "").to_string();

            commits.push_str(&format!(
                "[`{}`](<{}>) {}\n",
                &c.id[..7],
                c.url,
                clean_md_msg
            ));
        }

        let branch = self.ref_.strip_prefix("refs/heads/").unwrap_or(&self.ref_);
        let footer = format!(
            "\n- [{}](<{}>) on [{}](<{}>)/[{}](<{}/tree/{}>)",
            self.pusher.name,
            self.sender.html_url,
            self.repository.name,
            self.repository.html_url,
            branch,
            self.repository.html_url,
            branch,
        );

        let limit = 2000 - (footer.chars().count() + "...".len() + 1);
        if commits.chars().count() > limit {
            let mut truncated = commits.chars().take(limit).collect::<String>() + "...";
            if !truncated.ends_with(">)") {
                let lines: Vec<&str> = truncated.lines().collect();
                truncated = lines[..lines.len().saturating_sub(1)].join("\n");
            }
            commits = truncated + "\n";
        }

        Ok(Some(WebhookMessage {
            content: format!("{}{}", commits, footer),
            username: self.pusher.name.clone(),
            avatar_url: self.sender.avatar_url.clone(),
        }))
    }
}

File: server/handlers/push.go
package handlers

import (
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"webhook/structs/discord"
	"webhook/structs/github"
)

type push struct {
	Commits    []github.Commit   `json:"commits"`
	Ref        string            `json:"ref"`
	Sender     github.User       `json:"sender"`
	Pusher     github.User       `json:"pusher"`
	Repository github.Repository `json:"repository"`
}

func (e *push) handle(url.Values) (*discord.Webhook, error) {
	if len(e.Commits) == 0 {
		return nil, nil
	}

	linkRe, err := regexp.Compile(`\[([^]]+)]\((https?://[^)]+)\)`)
	if err != nil {
		return nil, err
	}
	mdRe, err := regexp.Compile(`(?m)^\s*#{1,3}\s+`)
	if err != nil {
		return nil, err
	}

	var commits strings.Builder
	for _, c := range e.Commits {
		lines := strings.Split(c.Message, "\n")
		first := ""
		if len(lines) > 0 {
			first = lines[0]
		}

		commitMsg := first
		if len(lines) > 1 {
			commitMsg += "..."
		}

		cleanLinkMsg := linkRe.ReplaceAllString(commitMsg, "[$1](<$2>)")
		cleanMdMsg := mdRe.ReplaceAllString(cleanLinkMsg, "")

		commits.WriteString(fmt.Sprintf("[`%s`](<%s>) %s\n", c.Id[:7], c.Url, cleanMdMsg))
	}

	branch := strings.TrimPrefix(e.Ref, "refs/heads/")
	footer := fmt.Sprintf(
		"\n- [%s](<%s>) on [%s](<%s>)/[%s](<%s/tree/%s>)",
		e.Pusher.Name,
		e.Sender.HtmlUrl,
		e.Repository.Name,
		e.Repository.HtmlUrl,
		branch,
		e.Repository.HtmlUrl,
		branch,
	)

	limit := 2000 - (len([]rune(footer)) + len("...") + 1)
	if len([]rune(commits.String())) > limit {
		truncated := string([]rune(commits.String())[:limit]) + "..."
		if !strings.HasSuffix(truncated, ">)") {
			lines := strings.Split(truncated, "\n")
			if len(lines) > 1 {
				truncated = strings.Join(lines[:len(lines)-1], "\n")
			}
		}
		commits.Reset()
		commits.WriteString(truncated + "\n")
	}

	return &discord.Webhook{
		Content:   commits.String() + footer,
		Username:  e.Pusher.Name,
		AvatarUrl: e.Sender.AvatarUrl,
	}, nil
}
